elements/pl-order-blocks/dag_checker.py

```diff
@@ -1,72 +1,60 @@
-            if group_id is None and answer_state.current_group is None:
-                answer_state.set_valid(block)
-            elif group_id is not None and answer_state.current_group is None:
-                answer_state.begin_group(group_id)
-                answer_state.set_valid(block)
-            elif group_id is None and answer_state.current_group is not None:
-                first_wrong = i
-                break
-            elif group_id is not None and answer_state.current_group is not None:
-                if group_id == answer_state.current_group:
-                    answer_state.set_valid(block)
-                    if answer_state.get_current_group_blocks() == group_sizes.get(group_id):
-                        answer_state.end_group()
-                else:
-                    first_wrong = i
-                    break
-        else:
-            first_wrong = i
-            break
-
-    return len(answer_state.valid), first_wrong
+from collections import Counter
+from typing import Mapping, Optional
+import networkx as nx  # type: ignore
+
+
+def check_topological_sorting(order: list[str], graph: nx.DiGraph) -> int:
+    seen = set()
+    for i, node in enumerate(order):
+        if node is None or not all(u in seen for (u, _) in graph.in_edges(node)):
+            return i
+        seen.add(node)
+    return len(order)
+
+
+def check_grouping(order: list[str], group_belonging: Mapping[str, Optional[int]]) -> int:
+    group_sizes = Counter(group_belonging.values())
+    cur_group = None
+    cur_group_size = None
+    for i, node in enumerate(order):
+        group_id = group_belonging.get(node)
+        if group_id is not None and cur_group is None:
+            cur_group = group_id
+            cur_group_size = 1
+        elif group_id is None and cur_group is not None:
+            return i
+        elif group_id is not None and cur_group is not None:
+            if group_id == cur_group:
+                cur_group_size += 1
+                if cur_group_size == group_sizes[cur_group]:
+                    cur_group = None
+                    cur_group_size = None
+            else:
+                return i
+    return len(order)
+
+
+def grade_dag(order: list[str], depends_graph: Mapping[str, list[str]], group_belonging: Mapping[str, Optional[int]]) -> tuple[int, int]:
+    """In order for a student submission to a DAG graded question to be deemed correct, the student
+    submission must be a topological sort of the DAG and blocks which are in the same pl-block-group
+    as one another must all appear contiguously.
+    :param order: the block ordering given by the student
+    :param depends_graph: The dependency graph between blocks specified in the question
+    :param group_belonging: which pl-block-group each block belongs to, specified in the question
+    :return: length of list that meets both correctness conditions, starting from the beginning,
+    index of first wrong (-1 if none wrong)
```

Comment:
```suggestion
    :param group_belonging: which pl-block-group each block belongs to, specified in the question
    :return: a tuple containing the length of list that meets both correctness conditions,
        starting from the beginning, and the index of first wrong (-1 if none wrong)
```

This is just a question (not necessarily asking for a code change), but is returning both values necessary? Seems like the caller could use a single value - if `ret == len(order)`, full credit, else use `ret` as the index of the first wrong.

Reply:
No, its definitely not necessary. The main reason that I kept it this way for now is that I wanted to keep this refactor contained to just this file. 

The return value here is definitely going to change when I implement partial credit (coming soon!) anyway.